In [37]:
%pip install kaggle

Note: you may need to restart the kernel to use updated packages.


In [38]:
!kaggle datasets download -d mlg-ulb/creditcardfraud -p ../data/raw --unzip

Dataset URL: https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud
License(s): DbCL-1.0




  0%|          | 0.00/66.0M [00:00<?, ?B/s]
  2%|▏         | 1.00M/66.0M [00:00<00:41, 1.62MB/s]
  3%|▎         | 2.00M/66.0M [00:00<00:27, 2.43MB/s]
  5%|▍         | 3.00M/66.0M [00:01<00:22, 2.89MB/s]
  6%|▌         | 4.00M/66.0M [00:01<00:20, 3.18MB/s]
  8%|▊         | 5.00M/66.0M [00:01<00:18, 3.38MB/s]
  9%|▉         | 6.00M/66.0M [00:02<00:18, 3.48MB/s]
 11%|█         | 7.00M/66.0M [00:02<00:17, 3.55MB/s]
 12%|█▏        | 8.00M/66.0M [00:02<00:16, 3.61MB/s]
 14%|█▎        | 9.00M/66.0M [00:02<00:16, 3.65MB/s]
 15%|█▌        | 10.0M/66.0M [00:03<00:16, 3.66MB/s]
 17%|█▋        | 11.0M/66.0M [00:03<00:15, 3.63MB/s]
 18%|█▊        | 12.0M/66.0M [00:03<00:15, 3.61MB/s]
 20%|█▉        | 13.0M/66.0M [00:04<00:15, 3.59MB/s]
 21%|██        | 14.0M/66.0M [00:04<00:15, 3.60MB/s]
 23%|██▎       | 15.0M/66.0M [00:04<00:14, 3.59MB/s]
 24%|██▍       | 16.0M/66.0M [00:04<00:14, 3.59MB/s]
 26%|██▌       | 17.0M/66.0M [00:05<00:14, 3.58MB/s]
 27%|██▋       | 18.0M/66.0M [00:05<00:14, 3.57MB/s]
 

In [39]:
import pandas as pd
import numpy as np
import kagglehub

df_raw = pd.read_csv("../data/raw/creditcard.csv")
df = df_raw.copy()

In [40]:
df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


In [41]:
df.shape

(284807, 31)

In [42]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='str')

In [43]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     284807 non-nu

In [44]:
df.describe()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
count,284807.000000,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,...,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,2.848070e+05,284807.000000,284807.000000
mean,94813.859575,1.168375e-15,3.416908e-16,-1.379537e-15,2.074095e-15,9.604066e-16,1.487313e-15,-5.556467e-16,1.213481e-16,-2.406331e-15,...,1.654067e-16,-3.568593e-16,2.578648e-16,4.473266e-15,5.340915e-16,1.683437e-15,-3.660091e-16,-1.227390e-16,88.349619,0.001727
std,47488.145955,1.958696e+00,1.651309e+00,1.516255e+00,1.415869e+00,1.380247e+00,1.332271e+00,1.237094e+00,1.194353e+00,1.098632e+00,...,7.345240e-01,7.257016e-01,6.244603e-01,6.056471e-01,5.212781e-01,4.822270e-01,4.036325e-01,3.300833e-01,250.120109,0.041527
min,0.000000,-5.640751e+01,-7.271573e+01,-4.832559e+01,-5.683171e+00,-1.137433e+02,-2.616051e+01,-4.355724e+01,-7.321672e+01,-1.343407e+01,...,-3.483038e+01,-1.093314e+01,-4.480774e+01,-2.836627e+00,-1.029540e+01,-2.604551e+00,-2.256568e+01,-1.543008e+01,0.000000,0.000000
25%,54201.500000,-9.203734e-01,-5.985499e-01,-8.903648e-01,-8.486401e-01,-6.915971e-01,-7.682956e-01,-5.540759e-01,-2.086297e-01,-6.430976e-01,...,-2.283949e-01,-5.423504e-01,-1.618463e-01,-3.545861e-01,-3.171451e-01,-3.269839e-01,-7.083953e-02,-5.295979e-02,5.600000,0.000000
50%,84692.000000,1.810880e-02,6.548556e-02,1.798463e-01,-1.984653e-02,-5.433583e-02,-2.741871e-01,4.010308e-02,2.235804e-02,-5.142873e-02,...,-2.945017e-02,6.781943e-03,-1.119293e-02,4.097606e-02,1.659350e-02,-5.213911e-02,1.342146e-03,1.124383e-02,22.000000,0.000000
75%,139320.500000,1.315642e+00,8.037239e-01,1.027196e+00,7.433413e-01,6.119264e-01,3.985649e-01,5.704361e-01,3.273459e-01,5.971390e-01,...,1.863772e-01,5.285536e-01,1.476421e-01,4.395266e-01,3.507156e-01,2.409522e-01,9.104512e-02,7.827995e-02,77.165000,0.000000
max,172792.000000,2.454930e+00,2.205773e+01,9.382558e+00,1.687534e+01,3.480167e+01,7.330163e+01,1.205895e+02,2.000721e+01,1.559499e+01,...,2.720284e+01,1.050309e+01,2.252841e+01,4.584549e+00,7.519589e+00,3.517346e+00,3.161220e+01,3.384781e+01,25691.160000,1.000000


In [45]:
df[['Time', 'Amount', 'Class']].head(10)

,Time,Amount,Class
0,0.0,149.62,0
1,0.0,2.69,0
2,1.0,378.66,0
3,1.0,123.50,0
4,2.0,69.99,0
5,2.0,3.67,0
6,4.0,4.99,0
7,7.0,40.80,0
8,7.0,93.20,0
9,9.0,3.68,0


In [46]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

In [47]:
df.isna().sum().sum()

0

In [48]:
df.duplicated(subset=df.columns.tolist()).sum()

1081

In [49]:
df.drop_duplicates(subset=df.columns.tolist(), inplace=True)

In [50]:
fraud_rate = df['Class'].mean() / 100
print(f"Fraud Rate: {fraud_rate:.6%}")

Fraud Rate: 0.001667%


In [51]:
import plotly.express as px
import matplotlib.pyplot as plt
import seaborn as sns

px.pie(df, names='Class', title='Fraud Distribution').show()

In [52]:
px.box(df, x='Class', y='Amount', title='Transaction Amount by Class').show()

In [53]:
px.line(df, x='Time', y='Amount', color='Class', title='Transaction Amount Over Time').show()

Fraudulent transactions generally spend way less than normal transactions (try to remain hidden) and only make up a very small fraction of all transactions (0.167%)

In [54]:
df.corr()['Class'].abs().sort_values(ascending=False)

Class     1.000000
V17       0.313498
V14       0.293375
V12       0.250711
V10       0.206971
V16       0.187186
V3        0.182322
V7        0.172347
V11       0.149067
V4        0.129326
V18       0.105340
V1        0.094486
V9        0.094021
V5        0.087812
V2        0.084624
V6        0.043915
V19       0.033631
V8        0.033068
V21       0.026357
V27       0.021892
V20       0.021486
Time      0.012359
V28       0.009682
V24       0.007210
V23       0.006333
Amount    0.005777
V22       0.004887
V26       0.004265
V13       0.003897
V15       0.003300
V25       0.003202
Name: Class, dtype: float64

In [55]:
px.imshow(df.corr(), text_auto=True, title='Correlation Matrix').show()

V features are pca components (vectors) and cannot be interpreted in a business setting

V17, V14, V12 are the highest three correlated features with Fraud

In [56]:
df = df.sort_values('Time')
train_end = int(len(df) * 0.70)
val_end = int(len(df) * 0.85)

train_df = df.iloc[:train_end]
val_df = df.iloc[train_end:val_end]
test_df = df.iloc[val_end:]

Time validation used to split data instead of random spliting to avoid leakage.

Oldest data used for training, and newer data used for validation and testing. 

In [57]:
train_df.to_csv("../data/processed/train.csv", index=False)
val_df.to_csv("../data/processed/val.csv", index=False)
test_df.to_csv("../data/processed/test.csv", index=False)